crates/bifrost/src/providers/replicated_loglet/read_path/read_stream_task.rs

```diff
@@ -0,0 +1,491 @@
+            permit.send(Ok(LogEntry::new_data(self.read_pointer, record)));
+            CacheReadResult::Sent
+        } else {
+            CacheReadResult::Miss
+        }
+    }
+
+    async fn readahead_from_server<T: TransportConnect>(
+        &self,
+        server: PlainNodeId,
+        to_offset: LogletOffset,
+        networking: &Networking<T>,
+        timeout: Duration,
+    ) -> Result<ServerReadResult, OperationError> {
+        let request = GetRecords {
+            header: LogServerRequestHeader::new(self.my_params.loglet_id, self.last_known_tail),
+            total_limit_in_bytes: None,
+            filter: self.filter.clone(),
+            from_offset: self.read_pointer,
+            to_offset,
+        };
+        trace!(
+            loglet_id = %self.my_params.loglet_id,
+            from_offset = %self.read_pointer,
+            %to_offset,
+            "Attempting to read records from {}",
+            server
+        );
+
+        let maybe_records = self
+            .logservers_rpc
+            .get_records
+            .call_timeout(networking, server, request, timeout)
+            .await;
+
+        match maybe_records {
+            Ok(records) => {
+                self.global_tail_watch
+                    .notify_offset_update(records.body().known_global_tail);
+                Ok(ServerReadResult::Records(records.into_body().records))
+            }
+            Err(NetworkError::Shutdown(e)) => Err(OperationError::Shutdown(e)),
+            Err(e) => {
+                trace!(
+                    loglet_id = %self.my_params.loglet_id,
+                    from_offset = %self.read_pointer,
+                    %to_offset,
+                    ?e,
+                    "Could not request record batch from node {}", server
+                );
+                Ok(ServerReadResult::Skip)
+            }
+        }
+    }
+}
+
+enum CacheReadResult {
+    /// Record was found but didn't match filter, read_pointer advanced
+    Filtered,
+    /// Record was found and sent
+    Sent,
+    /// Not in cache, read_pointer not advanced
+    Miss,
+    /// We should not read the next record (out of permits, or will exceed the last_known_tail)
+    Stop,
+}
+
+enum ServerReadResult {
+    /// Maybe got some records for you
+    Records(Vec<(LogletOffset, MaybeRecord)>),
+    /// Unreachable or failing node, skip and try another
+    Skip,
+}
+
+fn shuffle_nodeset_for_reads(nodeset: &NodeSet, my_node_id: PlainNodeId) -> Vec<PlainNodeId> {
+    // it's okay if it has one extra space if we are not in nodeset
+    let has_my_node_id = nodeset.contains(&my_node_id);
+    let mut new_nodeset: Vec<_> = nodeset.iter().cloned().collect();
+    // Shuffle nodes
+    new_nodeset.shuffle(&mut rand::thread_rng());
+    // put my node at the end if it's there
+    if has_my_node_id {
+        new_nodeset.retain(|&x| x != my_node_id);
+        new_nodeset.push(my_node_id);
+    }
+
+    new_nodeset
```

Comment:
Micro-optimization:

```
    let mut new_nodeset: Vec<_> = nodeset.iter().cloned().collect();
    // Shuffle nodes
    new_nodeset.shuffle(&mut rand::thread_rng());

    let has_my_node_idx = nodeset.iter().position(|&x| x == my_node_id);

    // put my node at the end if it's there
    if let Some(idx) = has_my_node_idx {
        new_nodeset.swap(idx, new_nodeset.len() - 1);
    }

    new_nodeset
```

could avoid some unnecessary moving of other elements.